CONTRIBUTING.md

```diff
@@ -0,0 +1,271 @@
+        Shortcuts defined here also have the advantage of having access to vscode's when clause contexts, and are easier
+        for users to override.
+    -   The keybindings are generated by scripts in `scripts/keybindings`. To update them, run `npm run keybind`.
+-   `vim/`: The entrypoint for any code run within nvim.
+    -   `vscode-neovim.vim`: The main vimscript file. This is where the communication between vscode and nvim is set up,
+    -   The remaining files override various nvim commands and bindings to forward them to vscode for handling.
+-   `runtime/lua`: The lua module that is loaded by nvim. Files in this directory are loaded by `vscode-neovim.vim`, and
+    serve the exact same purpose as the vimscript files. The long term goal is to replace all vimscript with lua.
+
+## How it works
+
+-   VScode connects to Neovim instance.
+-   When opening a file, a scratch buffer is created within Neovim and being initialized with text content from VSCode.
+-   Normal/visual mode commands are being sent directly to Neovim. The extension listens for buffer events and applies
+    edits from Neovim.
+-   When entering the insert mode, the extensions stops listen for keystroke events and delegates typing mode to VSCode.
+    Changes are synced to neovim in periodic intervals.
+-   After pressing escape key from the insert mode, extension sends changes obtained from the insert mode to Neovim.
+
+## Neovim APIs
+
+### Node Client
+
+VSCode-Neovim uses neovim's [`node-client`](https://neovim.io/node-client/index.html) library to communicate with
+neovim. This library is a wrapper around neovim's [msgpack-rpc API](https://neovim.io/doc/user/api.html). The library is
+used to send and receive messages to and from neovim. The library is also used to spawn neovim processes. Many functions
+are exposed, listed [here](https://neovim.io/node-client/classes/NeovimClient.html). However, the library is not kept up
+to date, so many API wrappers are missing. In these cases, the raw `call`, `callAtomic`, `lua`, and `command` commands
+are directly used.
+
+### Neovim UI Protocol
+
+VSCode-Neovim is built around neovim's ui protocol [:help ui](https://neovim.io/doc/user/ui.html). This protocol allows
+vscode to listen to events emitted by neovim, such as `mode_change`, `hl_attr_define`, and `win_viewport`. This API
+allows neovim to externalize the _rendering_ of various components of the editor, such as the cursor, highlights,
+buffers, cmdline, etc. When possible, this protocol should be used as the source of truth for the state of the editor.
+
+However, the UI protocol [is incomplete](https://github.com/neovim/neovim/issues/9421), and there are many holes in the
+API. For example, there is no way to get the type of visual mode or the current visual selection, `win_viewport` is not
+triggered on horizontal scrolling, there are no events for window splits and tabs, etc. As the nvim UI extension API is
```

Comment:
is there an upstream issue for the win_viewport horiz scrolling? https://github.com/neovim/neovim/issues?q=is%3Aissue+is%3Aopen+win_viewport

Reply:
@justinmk WinScrolled is currently used as a workaround, but the point here is to illustrate that there are gaps